src/DoctrineMessageRepository/DoctrineMessageRepository.php

```diff
@@ -0,0 +1,200 @@
+<?php
+
+namespace EventSauce\MessageRepository\DoctrineMessageRepository;
+
+use Doctrine\DBAL\Connection;
+use Doctrine\DBAL\Query\QueryBuilder;
+use EventSauce\IdEncoding\BinaryUuidIdEncoder;
+use EventSauce\IdEncoding\IdEncoder;
+use EventSauce\EventSourcing\AggregateRootId;
+use EventSauce\EventSourcing\Header;
+use EventSauce\EventSourcing\Message;
+use EventSauce\EventSourcing\MessageRepository;
+use EventSauce\EventSourcing\OffsetCursor;
+use EventSauce\EventSourcing\PaginationCursor;
+use EventSauce\EventSourcing\Serialization\MessageSerializer;
+use EventSauce\EventSourcing\UnableToPersistMessages;
+use EventSauce\EventSourcing\UnableToRetrieveMessages;
+use EventSauce\MessageRepository\TableSchema\DefaultTableSchema;
+use EventSauce\MessageRepository\TableSchema\TableSchema;
+use Generator;
+use LogicException;
+use Ramsey\Uuid\Uuid;
+use Throwable;
+use function array_keys;
+use function array_map;
+use function array_merge;
+use function count;
+use function get_class;
+use function implode;
+use function json_decode;
+use function json_encode;
+use function sprintf;
+
+class DoctrineMessageRepository implements MessageRepository
+{
+    private TableSchema $tableSchema;
+    private IdEncoder $aggregateIdEncoder;
+    private IdEncoder $eventIdEncoder;
+
+    public function __construct(
+        private Connection $connection,
+        private string $tableName,
+        private MessageSerializer $serializer,
+        private int $jsonEncodeOptions = 0,
+        ?TableSchema $tableSchema = null,
+        ?IdEncoder $aggregateRootIdEncoder = null,
+        ?IdEncoder $eventIdEncoder = null,
+    )
+    {
+        $this->tableSchema = $tableSchema ?? new DefaultTableSchema();
+        $this->aggregateIdEncoder = $aggregateRootIdEncoder ?? new BinaryUuidIdEncoder();
+        $this->eventIdEncoder = $eventIdEncoder ?? $this->aggregateIdEncoder;
```

Comment:
Should this be assumed? Thinking of my own use case here. I want a specific encoder for my aggregate root IDs, but I'm perfectly happy for my event IDs to be automatically generated and assigned UUIDs.

Should we assume that most people will want the same encoder for both? Or is defaulting both to `BinaryUuidIdEncoder` (independently) a more sensible option?

Reply:
My expectation is that most will use uuid/v4 and they'll want to use the same encoding for both. With that expectation, I think it makes sense. It can only be optimised for one case while still allowing the flexibility of more specific configurations.

Reply:
I understood what you meant. The case I'm optimising for is 1) the most likely ID format, and 2) when that format defaults to a serialisation strategy, use it for both. The least likely case is using your own serialiser, so that requires more explicit configuration, which I'm fine with. I understand your desire to change this, as it makes your specific case easier, I just don't believe it's the most common case, which is why I'm not optimising for it.

Reply:
Oh no, I have no particular desire to change this. That's a minor adjustment in my own code I can definitely live with. I was just using my own example as anecdotal evidence that maybe people are more likely to pick a different aggregate root ID type independently from that of event IDs.

But like I said, if you think that's the right assumption, go for it!

Reply:
👍